runtime/compiler/optimizer/BoolArrayStoreTransformer.cpp

```diff
@@ -0,0 +1,678 @@
+ *    type signature of the parm is used for telling the actual type.
+ */
+bool TR_BoolArrayStoreTransformer::isByteArrayNode(TR::Node *node, bool parmAsAuto)
+   {
+   if (parmAsAuto && node->getOpCode().hasSymbolReference() && node->getSymbolReference()->getSymbol()->isParm())
+      return false;
+   return getArrayDimension(node, false /*boolType*/) == 1;
+   }
+
+int TR_BoolArrayStoreTransformer::getArrayDimension(const char * signature, int length, bool boolType)
+   {
+   char expectedTypeChar = boolType? 'Z' : 'B';
+   return (signature && length >= 2 && signature[length-1] == expectedTypeChar && signature[length-2] == '[') ? length-1: -1;
+   }
+
+/*
+ * \brief
+ *     Get dimension of the array of given type
+ *
+ * \parm node
+ *    The node to look at
+ *
+ * \parm boolType
+ *    True if asking for boolean array and false if asking for byte array
+ *
+ * \return
+ *    The dimension of the array of given type. Return -1 if the node is not the
+ *    given type.
+ */
+int TR_BoolArrayStoreTransformer::getArrayDimension(TR::Node *node, bool boolType)
+   {
+   int nodeArrayDimension = -1;
+   if (node->getOpCodeValue() == TR::newarray)
+      {
+      int32_t expectedTypeValue = boolType ? 4: 8;
+      TR::Node *arrayTypeNode = node->getSecondChild();
+      TR_ASSERT(arrayTypeNode->getOpCode().isLoadConst(), "expect the second child of TR::newarray to be constant " POINTER_PRINTF_FORMAT, arrayTypeNode);
+      if (arrayTypeNode->getInt() == expectedTypeValue)
+         nodeArrayDimension = 1;
+      }
+   else
+      {
+      int length;
+      const char * signature = node->getTypeSignature(length);
+      nodeArrayDimension = getArrayDimension(signature, length, boolType);
+      }
+	return nodeArrayDimension;
+   }
+
+/*
+ * \brief
+ *    Find all the loads of autos or parms in a subtree and figure out its type
+ *    first time the load is referenced.
+ *
+ * \parm node
+ *    The subtree to look at
+ *
+ * \parm typeInfo
+ *    The type infomation of each auto at the current subtree
+ *
+ * \parm boolArrayNodes
+ *    Load of autos or parms that are [Z
+ *
+ * \parm byteArrayNodes
+ *    Load of autos or parms that are [B
+ *
+ * \parm loadAutoNodes
+ *    All the loads of autos or parms in the containing block. An address load is added to this list the first time
+ *    seen in the trees
+ */
+void TR_BoolArrayStoreTransformer::findLoadAddressAutoAndFigureOutType(TR::Node *node, TypeInfo * typeInfo, TR::NodeChecklist &boolArrayNodes, TR::NodeChecklist &byteArrayNodes, TR::NodeChecklist &loadAutoNodes)
+   {
+   for (int i = 0; i < node->getNumChildren(); i++)
+      findLoadAddressAutoAndFigureOutType(node->getChild(i), typeInfo, boolArrayNodes, byteArrayNodes, loadAutoNodes);
+
+   if (node->getType() == TR::Address && node->getOpCode().isLoadDirect() && node->getOpCode().hasSymbolReference() &&
+         node->getSymbolReference()->getSymbol()->isAutoOrParm() && !loadAutoNodes.contains(node))
+      {
```

Comment:
To avoid taking exponential time, `loadAutoNodes` should probably apply to all nodes encountered by this method rather than only the address-typed auto loads.

Reply:
Fixed according to comments. Thanks!